exportoptionsgenerator/exportoptionsgenerator.go

```diff
@@ -21,128 +21,118 @@ const (
+func NewWithInfoProvider(infoProvider InfoProvider, xcodeVersionReader xcodeversion.Reader, logger log.Logger) ExportOptionsGenerator {
+	return ExportOptionsGenerator{
+		xcodeVersionReader:  xcodeVersionReader,
+		certificateProvider: LocalCodesignIdentityProvider{},
+		profileProvider:     LocalProvisioningProfileProvider{},
+		archiveInfoProvider: infoProvider,
+		logger:              logger,
+	}
 }
 
 // GenerateApplicationExportOptions generates exportOptions for an application export.
-func (g ExportOptionsGenerator) GenerateApplicationExportOptions(
-	exportMethod exportoptions.Method,
-	containerEnvironment string,
-	teamID string,
-	uploadBitcode bool,
-	compileBitcode bool,
-	archivedWithXcodeManagedProfiles bool,
-	codeSigningStyle exportoptions.SigningStyle,
-	testFlightInternalTestingOnly bool,
-) (exportoptions.ExportOptions, error) {
+func (g ExportOptionsGenerator) GenerateApplicationExportOptions(exportMethod exportoptions.Method, codeSigningStyle exportoptions.SigningStyle, opts Opts) (exportoptions.ExportOptions, error) {
 	xcodeVersion, err := g.xcodeVersionReader.GetVersion()
 	if err != nil {
 		return nil, fmt.Errorf("failed to get Xcode version: %w", err)
 	}
 
-	mainTargetBundleID, entitlementsByBundleID, err := g.applicationTargetsAndEntitlements(exportMethod)
+	archiveInfo, err := g.archiveInfoProvider.Read()
 	if err != nil {
 		return nil, err
 	}
 
-	iCloudContainerEnvironment, err := determineIcloudContainerEnvironment(containerEnvironment, entitlementsByBundleID, exportMethod, xcodeVersion.Major)
+	// app-store-connect exports should contain App Clip too, but not other export methods:
+	// ..
+	// <key>provisioningProfiles</key>
+	// <dict>
+	// 	<key>io.bundle.id</key>
+	// 	<string>Development Application Profile</string>
+	// 	<key>io.bundle.id.AppClipID</key>
+	// 	<string>Development App Clip Profile</string>
+	// </dict>
+	// ..
+	for bundleID := range archiveInfo.EntitlementsByBundleID {
+		if bundleID == archiveInfo.AppClipBundleID {
+			if !exportMethod.IsAppStore() && bundleID != archiveInfo.MainBundleID { // do not remove if exporting App Clip
+				g.logger.Debugf("Filtering out App Clip target: %s", bundleID)
+				delete(archiveInfo.EntitlementsByBundleID, bundleID)
+			}
+		}
+	}
+
+	iCloudContainerEnvironment, err := determineIcloudContainerEnvironment(opts.ContainerEnvironment, archiveInfo.EntitlementsByBundleID, exportMethod, xcodeVersion.Major)
 	if err != nil {
 		return nil, err
 	}
 
-	exportOpts := generateBaseExportOptions(exportMethod, xcodeVersion, uploadBitcode, compileBitcode, iCloudContainerEnvironment)
+	exportOpts := generateBaseExportOptions(exportMethod, xcodeVersion, opts.UploadBitcode, opts.CompileBitcode, iCloudContainerEnvironment)
 
 	if xcodeVersion.Major >= 12 {
-		exportOpts = addDistributionBundleIdentifierFromXcode12(exportOpts, mainTargetBundleID)
+		exportOpts = addDistributionBundleIdentifierFromXcode12(exportOpts, archiveInfo.MainBundleID)
```

Comment:
I originally wanted to leave a comment, that distributionBundleID only needs to be added for non app store methods, but as I see this is already implemented inside the `addDistributionBundleIdentifierFromXcode12` function.

Reply:
Also, I suggest renaming archiveInfo.MainBundleID to rchiveInfo.ProductToDistributeBundleID or something similar, which explains that you either export the main app target or the app clip target.

Reply:
Renamed MainBundleID.